### ISSUE TITLE
git push fails on GitHub private repository

### ISSUE THREAD

**Issue by delcypher** [git-lfs/git-lfs#1878]
This is my first time using `git-lfs` so apologies if I get some of the details wrong.

I installed the `git-lfs` client from the [AUR](https://aur.archlinux.org/packages/git-lfs/).

I did the following

```
git lfs install
git lfs track "*.tar.xz"
git add myfile.tar.xz
git commit -m "test"
git push
```

The push fails with a very unhelpful error message (note the real repository name has been replaced with `secret-repo`).

```
locking: http: http: Not Found
Docs: https://github.com/contact
error: failed to push some refs to 'git@github.com:delcypher/secret-repo.git'
```

If I set the `GIT_TRACE=1` environment variable I see more details

```
GIT_TRACE=1 git push origin
17:51:34.420744 git.c:371               trace: built-in: git 'push' 'origin'
17:51:34.421450 run-command.c:350       trace: run_command: 'ssh' 'git@github.com' 'git-receive-pack '\''delcypher/secret-repo.git'\'''
17:51:35.545179 run-command.c:350       trace: run_command: '.git/hooks/pre-push' 'origin' 'git@github.com:delcypher/secret-repo.git'
17:51:35.553467 git.c:600               trace: exec: 'git-lfs' 'pre-push' 'origin' 'git@github.com:delcypher/secret-repo.git'
17:51:35.553532 run-command.c:350       trace: run_command: 'git-lfs' 'pre-push' 'origin' 'git@github.com:delcypher/secret-repo.git'
trace git-lfs: run_command: 'git' version
trace git-lfs: run_command: 'git' config -l
trace git-lfs: tq: running as batched queue, batch size of 100
trace git-lfs: ssh: git@github.com git-lfs-authenticate delcypher/secret-repo.git download
trace git-lfs: HTTP: GET https://api.github.com/lfs/delcypher/secret-repo/locks
trace git-lfs: HTTP: 404
trace git-lfs: HTTP: {"message":"Not Found","documentation_url":"https://github.com/contact"}
locking: http: http: Not Found
Docs: https://github.com/contact
error: failed to push some refs to 'git@github.com:delcypher/secret-repo.git'
```

It looks like `git-lfs` is trying to do a HTTP GET on `https://api.github.com/lfs/delcypher/klee-comparison-misc/locks`. The repository is a private repository so `{"message":"Not Found","documentation_url":"https://github.com/contact"}` response probably means the GET is not being correctly authenticated.

It looks like the invocation of SSH is to get authentication information. If I run this I seem to get a sensible response (the real token has been replaced with `sometoken` in the output below).

```
$ ssh git@github.com git-lfs-authenticate delcypher/secret-repo.git download
{
  "href": "https://api.github.com/lfs/delcypher/secret-repo",
  "header": {
    "Authorization": "RemoteAuth sometoken"
  },
  "expires_at": "2017-01-22T18:12:10Z"
}
```

Here is the output of `git lfs env`

```
git lfs env
git-lfs/1.5.0 (GitHub; linux amd64; go 1.7.4)
git version 2.11.0

Endpoint=https://github.com/delcypher/secret-repo.git/info/lfs (auth=none)
  SSH=git@github.com:delcypher/secret-repo.git
LocalWorkingDir=/home/dsl11/dev/klee-afr/secret-repo
LocalGitDir=/home/dsl11/dev/klee-afr/secret-repo/.git
LocalGitStorageDir=/home/dsl11/dev/klee-afr/secret-repo/.git
LocalMediaDir=/home/dsl11/dev/klee-afr/secret-repo/.git/lfs/objects
LocalReferenceDir=
TempDir=/home/dsl11/dev/klee-afr/secret-repo/.git/lfs/tmp
ConcurrentTransfers=3
TusTransfers=false
BasicTransfersOnly=false
BatchTransfer=true
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic
UploadTransfers=basic
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

**Comment by delcypher** [comment]
I just tried using the `https` remote and that also failed :(

```
$ GIT_TRACE=1 git push https_origin master
18:34:36.856820 git.c:371               trace: built-in: git 'push' 'https_origin' 'master'
18:34:36.857676 run-command.c:350       trace: run_command: 'git-remote-https' 'https_origin' 'https://github.com/delcypher/secret-repo.git'
Username for 'https://github.com': delcypher
Password for 'https://delcypher@github.com':
18:34:41.309320 run-command.c:350       trace: run_command: '.git/hooks/pre-push' 'https_origin' 'https://github.com/delcypher/secret-repo.git'
18:34:41.317959 git.c:600               trace: exec: 'git-lfs' 'pre-push' 'https_origin' 'https://github.com/delcypher/secret-repo.git'
18:34:41.318033 run-command.c:350       trace: run_command: 'git-lfs' 'pre-push' 'https_origin' 'https://github.com/delcypher/secret-repo.git'
trace git-lfs: run_command: 'git' version
trace git-lfs: run_command: 'git' config -l
trace git-lfs: tq: running as batched queue, batch size of 100
trace git-lfs: HTTP: GET https://github.com/delcypher/secret-repo.git/info/lfs/locks
trace git-lfs: HTTP: 404
trace git-lfs: HTTP: {"message":"Not Found","documentation_url":"https://github.com/contact"}
locking: http: http: Not Found
Docs: https://github.com/contact
error: failed to push some refs to 'https://github.com/delcypher/secret-repo.git'
```

**Comment by delcypher** [comment]
Hmm. I've dug a little more into this and it seems that the Arch Linux AUR package I was using is built incorrectly. It's supposed to be version `1.5.5` but because of the way the packager builds it we end up building whatever is in the git-lfs master branch which reports its versions as `1.5.0`.

It seems they run `go get` and it ends up fetching what is in the `git-lfs` master branch.

I made a [hacked version of the PKGBUILD](https://gist.github.com/delcypher/120b6222f39d00539fee4c0229582a11) that builds and installs the right version (I think). With that version my `git push` works fine.

I'm not familiar enough with the correct build process for `git-lfs` to know the right way to fix the build.

**Comment by technoweenie** [comment]
The only "supported" official install we have for Arch Linux is by downloading the appropriate linux binary from [the LFS release page](https://github.com/git-lfs/git-lfs/releases/tag/v1.5.5). If you want to build it yourself, you'll have to change to the desired git tag in the repository.

I did notice this part from your PKGBUILD:

```
  GOPATH="$srcdir" go get -v -d
  # HACK: go get seems to put the master branch sources for git-lfs in
  # `$srcdir/src/github.com/git-lfs/git-lfs/`. I don't know why it does this.
  # so remove them and link in the right sources.
  rm -rf "$srcdir"/src/github.com/git-lfs/git-lfs
```

The `go get` command downloads a go package into the GOPATH. Since you don't want that, you can probably just remove those lines, and skip directly to the `ln -sf "$srcdir/$pkgname-$pkgver" "$srcdir"/src/github.com/git-lfs/git-lfs` line.

**Comment by technoweenie** [comment]
Oh yeah, and the actual bug here. It has to do with #1815. It has to support 404s, since no LFS server currently has a locking service. It's not ready for release of course, but we'll get some tweaks in this week so that master is still usable. Thanks for the report!

**Comment by delcypher** [comment]
@technoweenie Thanks for the informative replies. I was going to close this issue but it looks like @ttaylorr has been assigned to this so I'll leave it open.

**Comment by technoweenie** [comment]
Fixed in #1885. If someone tries to push with a build from master now, they'll just get a warning:

```sh
$ git push origin something/something/darkside -u
WARNING: Unable to search for locks contained in this push.
         Temporarily skipping check ...
Counting objects: 11, done.
Compressing objects: 100% (11/11), done.
...
```

Note that work towards LFS v2.0 is still going on. We're discussing more ideas to handle the pre-push hook better in https://github.com/git-lfs/git-lfs/issues/1884. I'm closing this now since the immediate bug blocking pushes is fixed though.

Thanks for the report!
